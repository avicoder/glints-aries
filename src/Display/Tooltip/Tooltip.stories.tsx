import React from 'react';
import { Story, Meta } from '@storybook/react';
import styled from 'styled-components';

import { Tooltip, StringTooltip, HtmlTooltip } from './Tooltip';
import { BaseContainer } from '../../Layout/GlintsContainer/GlintsContainer';
import Tag from '../../General/Tag/BasicTag';

const Container = styled(BaseContainer)`
  width: 100%;
  height: 150px;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const defaultText = 'Lorem Ipsum';

export default {
  title: 'General/Tooltip',
  component: Tooltip,
  decorators: [Story => <Container>{Story()}</Container>],
} as Meta;

const StringTooltipTemplate: Story<StringTooltip> = ({ text, ...args }) => (
  <Tooltip {...args} text={text || defaultText}>
    <Tag>Glints Aries</Tag>
  </Tooltip>
);

const HtmlTooltipTemplate: Story<HtmlTooltip> = ({ text, ...args }) => (
  <Tooltip {...args} text={text || defaultText}>
    <Tag>Glints Aries</Tag>
  </Tooltip>
);

export const Interactive = StringTooltipTemplate.bind({});
Interactive.args = {
  text: defaultText,
  position: 'top',
};

export const DifferentPosition = StringTooltipTemplate.bind({});
DifferentPosition.args = {
  position: 'left',
};

export const WithHtmlContent = HtmlTooltipTemplate.bind({});
WithHtmlContent.args = {
  text: <b>Hello World</b>,
  textAsString: 'Hello World',
};
