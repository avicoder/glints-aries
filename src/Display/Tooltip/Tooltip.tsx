import * as React from 'react';

import classNames from 'classnames';

import {
  TooltipContainer,
  TooltipContent,
  TooltipMessage,
} from './TooltipStyle';

const Tooltip: React.FunctionComponent<Props> = ({
  classes = {},
  children,
  text,
  position,
  ...defaultProps
}) => {
  const contentRef = React.useRef(null);
  const [isShow, setIsShow] = React.useState(false);
  const showTooltip = () => setIsShow(true);
  const hideTooltip = () => setIsShow(false);

  const hideTooltipIfTouchOutside = React.useCallback(
    (event: TouchEvent | React.TouchEvent) => {
      const hasTouchedOutsideOfTooltipContent = !contentRef.current.contains(
        event.target as HTMLElement
      );

      if (hasTouchedOutsideOfTooltipContent) {
        hideTooltip();
        document.removeEventListener('touchstart', hideTooltipIfTouchOutside);
      }
    },
    []
  );

  const handleTouchStart = () => {
    if (!isShow) {
      showTooltip();
      document.addEventListener('touchstart', hideTooltipIfTouchOutside);
    }
  };

  return (
    <TooltipContainer
      className={classNames('aries-tooltip', classes.container)}
      role="tooltip"
      aria-hidden={isShow ? 'false' : 'true'}
      aria-label={text}
      // The tooltip does not close on iOS devices because of this issue https://developer.mozilla.org/en-US/docs/Web/API/Element/click_event#Safari_Mobile
      // Adding onTouchStart and onTouchEnd as a workaround
      // On mobile, it shows the tooltip on touch and hides the tooltip when the touch is released
      onTouchStart={handleTouchStart}
      onMouseOver={showTooltip}
      onMouseLeave={hideTooltip}
      {...defaultProps}
    >
      {isShow && (
        <TooltipContent
          ref={contentRef}
          className={classNames('aries-tooltip-content', classes.content)}
          text={text}
          position={position}
        >
          <TooltipMessage
            className={classNames('aries-tooltip-message', classes.message)}
          >
            {text}
          </TooltipMessage>
        </TooltipContent>
      )}
      {children}
    </TooltipContainer>
  );
};

interface Classes {
  container?: string;
  content?: string;
  message?: string;
}

interface Props
  extends React.ComponentPropsWithoutRef<typeof TooltipContainer> {
  classes?: Classes;
  children: React.ReactNode;
  text: string;
  position?: string;
}

export default Tooltip;
