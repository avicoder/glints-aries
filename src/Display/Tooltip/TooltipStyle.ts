import styled from 'styled-components';
import { Greyscale } from '../../Utils/Colors';
import { Props } from './Tooltip';

export const TooltipContainer = styled.div`
  position: relative;
  display: inline-flex;
`;

export const TooltipContent = styled.div<Props>`
  position: absolute;
  z-index: 10;
  width: 100%;

  ${({ position }) => {
    switch (position) {
      case 'bottom':
        return `
          top: 100%;
          left: 50%;
          padding-top: 9px;
          transform: translateX(-50%);

          &:after {
            border-left: 9px solid transparent;
            border-right: 9px solid transparent;
            border-bottom: 9px solid ${Greyscale.black};
            top: 0;
            left: 50%;
            transform: translateX(-50%);
          }
        `;
      case 'left':
        return `
          top: 50%;
          right: 100%;
          padding-right: 9px;
          transform: translateY(-50%);
          
          &:after {
            border-left: 9px solid ${Greyscale.black};
            border-top: 9px solid transparent;
            border-bottom: 9px solid transparent;
            top: 50%;
            right: 0;
            transform: translateY(-50%);
          }
        `;
      case 'right':
        return `
          top: 50%;
          left: 100%;
          padding-left: 9px;
          transform: translateY(-50%);
          
          &::after {
            border-right: 9px solid ${Greyscale.black};
            border-top: 9px solid transparent;
            border-bottom: 9px solid transparent;
            top: 50%;
            left: 0;
            transform: translateY(-50%);
          }
        `;
      default:
        return `
          bottom: 100%;
          left: 50%;
          padding-bottom: 9px;
          transform: translateX(-50%);

          &:after {
            border-left: 9px solid transparent;
            border-right: 9px solid transparent;
            border-top: 9px solid ${Greyscale.black};
            bottom: 0;
            left: 50%;
            transform: translateX(-50%);
          }
        `;
    }
  }}

  &:after {
    content: '';
    position: absolute;
  }
`;

export const TooltipMessage = styled.div`
  background: ${Greyscale.black};
  border-radius: 3px;
  color: ${Greyscale.white};
  line-height: 1.5;
  font-size: 12px;
  padding: 10px;
  text-align: center;
`;
