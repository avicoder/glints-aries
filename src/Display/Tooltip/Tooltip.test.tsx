import * as React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

import Tooltip, { Position } from './Tooltip';

describe('<Tooltip/> snapshot', () => {
  test('should match snapshot when Tooltip is hidden', () => {
    const { asFragment } = render(
      <Tooltip
        text="mock text"
        position="right"
        classes={{
          container: 'container',
          content: 'content',
          message: 'message',
        }}
      >
        <span>mock content</span>
      </Tooltip>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<Tooltip/> snapshots with position prop', () => {
  const matchSnapshotWithPosition = (position: Position) => {
    test(`position ${position}`, () => {
      const { asFragment, queryByText } = render(
        <Tooltip
          text="mock text"
          position={position}
          classes={{
            container: 'container',
            content: 'content',
            message: 'message',
          }}
        >
          <span>mock content</span>
        </Tooltip>
      );
      const element = queryByText('mock content');
      fireEvent.mouseOver(element);
      expect(asFragment()).toMatchSnapshot();
    });
  };

  ['left', 'right', 'bottom', 'top'].forEach((position: Position) =>
    matchSnapshotWithPosition(position)
  );
});

describe('<Tooltip/> touch event', () => {
  const TooltipComponent = () => (
    <React.Fragment>
      <div>touch outside</div>
      <Tooltip text="mock text">
        <span>mock content</span>
      </Tooltip>
    </React.Fragment>
  );

  test('touchstart event should show and hide Tooltip', () => {
    const { queryByText } = render(<TooltipComponent />);
    expect(queryByText('mock text')).not.toBeInTheDocument();

    const element = queryByText('mock content');
    fireEvent.touchStart(element);
    expect(queryByText('mock text')).toBeInTheDocument();

    fireEvent.touchStart(element);
    expect(queryByText('mock text')).not.toBeInTheDocument();
  });

  test('touch outside of Tooltip should hide Tooltip', () => {
    const { queryByText } = render(<TooltipComponent />);
    expect(queryByText('mock text')).not.toBeInTheDocument();

    const element = queryByText('mock content');
    fireEvent.touchStart(element);
    expect(queryByText('mock text')).toBeInTheDocument();

    const outside = queryByText('touch outside');
    fireEvent.touchStart(outside);
    expect(queryByText('mock text')).not.toBeInTheDocument();
  });

  test('touch outside should register and unregister hideTooltipIfTouchOutside callback', () => {
    const addEventListenerSpy = jest.spyOn(document, 'addEventListener');
    const removeEventListenerSpy = jest.spyOn(document, 'removeEventListener');
    const useCallbackSpy = jest.spyOn(React, 'useCallback');

    const { queryByText } = render(<TooltipComponent />);
    const element = queryByText('mock content');
    fireEvent.touchStart(element);
    expect(addEventListenerSpy.mock.calls[0][1]).toEqual(
      useCallbackSpy.mock.results[0].value
    );

    const outside = queryByText('touch outside');
    fireEvent.touchStart(outside);

    expect(removeEventListenerSpy.mock.calls[0][1]).toEqual(
      useCallbackSpy.mock.results[0].value
    );

    jest.clearAllMocks();
  });
});
