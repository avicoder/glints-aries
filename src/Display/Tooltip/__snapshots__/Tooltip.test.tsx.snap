// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip/> snapshot should match snapshot when Tooltip is hidden 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

<div
    aria-hidden="true"
    aria-label="mock text"
    class="c0 aries-tooltip container"
    role="tooltip"
  >
    <span>
      mock content
    </span>
  </div>
</DocumentFragment>
`;

exports[`<Tooltip/> snapshot with html content should display the tooltip 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 {
  position: absolute;
  z-index: 10;
  width: 100%;
  top: 50%;
  left: 100%;
  padding-left: 9px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1::after {
  border-right: 9px solid #000000;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: 50%;
  left: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1:after {
  content: '';
  position: absolute;
}

.c2 {
  background: #000000;
  border-radius: 3px;
  color: #FFFFFF;
  line-height: 1.5;
  font-size: 12px;
  padding: 10px;
  text-align: center;
}

<div
    aria-hidden="false"
    aria-label="Hello World"
    class="c0 aries-tooltip container"
    role="tooltip"
  >
    <div
      class="c1 aries-tooltip-content content"
    >
      <div
        class="c2 aries-tooltip-message message"
      >
        <span>
          Hello World
        </span>
      </div>
    </div>
    <span>
      mock content
    </span>
  </div>
</DocumentFragment>
`;

exports[`<Tooltip/> snapshots with position prop position bottom 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 {
  position: absolute;
  z-index: 10;
  width: 100%;
  top: 100%;
  left: 50%;
  padding-top: 9px;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c1:after {
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  border-bottom: 9px solid #000000;
  top: 0;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c1:after {
  content: '';
  position: absolute;
}

.c2 {
  background: #000000;
  border-radius: 3px;
  color: #FFFFFF;
  line-height: 1.5;
  font-size: 12px;
  padding: 10px;
  text-align: center;
}

<div
    aria-hidden="false"
    aria-label="mock text"
    class="c0 aries-tooltip container"
    role="tooltip"
  >
    <div
      class="c1 aries-tooltip-content content"
    >
      <div
        class="c2 aries-tooltip-message message"
      >
        mock text
      </div>
    </div>
    <span>
      mock content
    </span>
  </div>
</DocumentFragment>
`;

exports[`<Tooltip/> snapshots with position prop position left 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 {
  position: absolute;
  z-index: 10;
  width: 100%;
  top: 50%;
  right: 100%;
  padding-right: 9px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1:after {
  border-left: 9px solid #000000;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: 50%;
  right: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1:after {
  content: '';
  position: absolute;
}

.c2 {
  background: #000000;
  border-radius: 3px;
  color: #FFFFFF;
  line-height: 1.5;
  font-size: 12px;
  padding: 10px;
  text-align: center;
}

<div
    aria-hidden="false"
    aria-label="mock text"
    class="c0 aries-tooltip container"
    role="tooltip"
  >
    <div
      class="c1 aries-tooltip-content content"
    >
      <div
        class="c2 aries-tooltip-message message"
      >
        mock text
      </div>
    </div>
    <span>
      mock content
    </span>
  </div>
</DocumentFragment>
`;

exports[`<Tooltip/> snapshots with position prop position right 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 {
  position: absolute;
  z-index: 10;
  width: 100%;
  top: 50%;
  left: 100%;
  padding-left: 9px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1::after {
  border-right: 9px solid #000000;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: 50%;
  left: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1:after {
  content: '';
  position: absolute;
}

.c2 {
  background: #000000;
  border-radius: 3px;
  color: #FFFFFF;
  line-height: 1.5;
  font-size: 12px;
  padding: 10px;
  text-align: center;
}

<div
    aria-hidden="false"
    aria-label="mock text"
    class="c0 aries-tooltip container"
    role="tooltip"
  >
    <div
      class="c1 aries-tooltip-content content"
    >
      <div
        class="c2 aries-tooltip-message message"
      >
        mock text
      </div>
    </div>
    <span>
      mock content
    </span>
  </div>
</DocumentFragment>
`;

exports[`<Tooltip/> snapshots with position prop position top 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 {
  position: absolute;
  z-index: 10;
  width: 100%;
  bottom: 100%;
  left: 50%;
  padding-bottom: 9px;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c1:after {
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  border-top: 9px solid #000000;
  bottom: 0;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c1:after {
  content: '';
  position: absolute;
}

.c2 {
  background: #000000;
  border-radius: 3px;
  color: #FFFFFF;
  line-height: 1.5;
  font-size: 12px;
  padding: 10px;
  text-align: center;
}

<div
    aria-hidden="false"
    aria-label="mock text"
    class="c0 aries-tooltip container"
    role="tooltip"
  >
    <div
      class="c1 aries-tooltip-content content"
    >
      <div
        class="c2 aries-tooltip-message message"
      >
        mock text
      </div>
    </div>
    <span>
      mock content
    </span>
  </div>
</DocumentFragment>
`;
